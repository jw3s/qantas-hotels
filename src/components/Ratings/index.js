import React from 'react';
import styled from 'styled-components';

const RatingWrapper = styled.div`
  width: 25px;
  height: 25px;
  clip-path: ${({ type }) => (type === 'star'
    ? 'polygon(50% 0%, 61% 35%, 98% 35%, 68% 57%, 79% 91%, 50% 70%, 21% 91%, 32% 57%, 2% 35%, 39% 35%)'
    : 'circle(10px at center)')};
  background: #f7f7f7;
  cursor: pointer;

  > div {
    width: ${({ value }) => value}%;
    height: 25px;
    background: #f6e200;
  }
`;

const Rating = ({ value, type }: { value: number, type: string }) => (
  <RatingWrapper value={value} type={type}>
    <div />
  </RatingWrapper>
);

const RatingsWrapper = styled.div`
  display: flex;
`;

const getPercentageValue = (val: number, index: number): number => {
  if (index <= val) return 100;
  if (index > val + 1) return 0;
  return (val % 1) * 100;
};

const Ratings = ({ value: rating, type = 'star', ...rest }: { value: number, type?: 'star' | 'circle' }) => {
  const stars = [];

  for (let i = 1; i <= 5; i += 1) {
    stars.push(
      <Rating type={type} {...rest} value={getPercentageValue(rating, i)} key={`${i}`} index={i} />,
    );
  }

  return <RatingsWrapper>{stars}</RatingsWrapper>;
};
Ratings.defaultProps = {
  type: 'star',
};

export default Ratings;
