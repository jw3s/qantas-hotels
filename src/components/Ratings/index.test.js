import React from 'react';
import renderer from 'react-test-renderer';

import Ratings from '.';

import 'jest-styled-components';

describe('<Ratings />', () => {
  it('<Ratings /> should render tree with type star', () => {
    const tree = renderer.create(<Ratings value={5} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('<Ratings /> should render tree with type circle', () => {
    const tree = renderer.create(<Ratings value={5} type="circle" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
