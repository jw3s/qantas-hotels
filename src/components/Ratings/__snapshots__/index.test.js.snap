// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ratings /> <Ratings /> should render tree with type circle 1`] = `
.c1 {
  width: 25px;
  height: 25px;
  -webkit-clip-path: circle(10px at center);
  clip-path: circle(10px at center);
  background: #f7f7f7;
  cursor: pointer;
}

.c1 > div {
  width: 100%;
  height: 25px;
  background: #f6e200;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
>
  <div
    className="c1"
    type="circle"
    value={100}
  >
    <div />
  </div>
  <div
    className="c1"
    type="circle"
    value={100}
  >
    <div />
  </div>
  <div
    className="c1"
    type="circle"
    value={100}
  >
    <div />
  </div>
  <div
    className="c1"
    type="circle"
    value={100}
  >
    <div />
  </div>
  <div
    className="c1"
    type="circle"
    value={100}
  >
    <div />
  </div>
</div>
`;

exports[`<Ratings /> <Ratings /> should render tree with type star 1`] = `
.c1 {
  width: 25px;
  height: 25px;
  -webkit-clip-path: polygon(50% 0%,61% 35%,98% 35%,68% 57%,79% 91%,50% 70%,21% 91%,32% 57%,2% 35%,39% 35%);
  clip-path: polygon(50% 0%,61% 35%,98% 35%,68% 57%,79% 91%,50% 70%,21% 91%,32% 57%,2% 35%,39% 35%);
  background: #f7f7f7;
  cursor: pointer;
}

.c1 > div {
  width: 100%;
  height: 25px;
  background: #f6e200;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
>
  <div
    className="c1"
    type="star"
    value={100}
  >
    <div />
  </div>
  <div
    className="c1"
    type="star"
    value={100}
  >
    <div />
  </div>
  <div
    className="c1"
    type="star"
    value={100}
  >
    <div />
  </div>
  <div
    className="c1"
    type="star"
    value={100}
  >
    <div />
  </div>
  <div
    className="c1"
    type="star"
    value={100}
  >
    <div />
  </div>
</div>
`;
